core/network/src/sync.rs

```diff
@@ -452,6 +463,46 @@ impl<B: BlockT> ChainSync<B> {
 		}
 	}
 
+	fn handle_ancestor_search_state(
+		state: AncestorSearchState<B>,
+		n: NumberFor<B>,
+		block_hash_match: bool,
+	) -> Option<(AncestorSearchState<B>, NumberFor<B>)> {
+		println!("ancestor search: {:?} {:?} {:?}", n, block_hash_match, state);
+		match state {
+			AncestorSearchState::ExponentialBackoff(next_distance_to_tip) => {
+				if block_hash_match && next_distance_to_tip == As::sa(1) {
+					// We found the ancestor so there is no more ancestor search state.
+					println!("ancestor found at {:?}", n);
+					return None;
```

Comment:
When is this case triggered? It seems that we only double `next_distance_to_tip`, so I guess this can be hit only on first run?

Reply:
Yes, only can be hit by first run.

Reply:
Maybe extend the comment to say that this can only happen in the first step and hence we don't need to binary search forward.